### PR TITLE
ADC driver report back vref alone with sample count

### PR DESCRIPTION
**Pull request by SalimTerryLi**
**Describe problem solved by this pull request**
Current ADC system does not provide reference voltage in any way. This PR focuses on introduce the v_ref in to ADC, potentially improve the ADC accuracy.

**Describe your solution**
I'm adding this new arch side function `float px4_arch_adc_reference_v(unsigned channel)` to get vref properly provided.
This function should be implemented differently on each board. On the side of stm32, this function should sample the internal 1.2V reference and apply factory calibration to get the true vref level.

**Describe possible alternatives**
In short term a fixed value can be set to calculate ADC results, but dynamically calculated value helps improving ADC accuracy.
vref would be provided within `px4_arch_adc_sample`.

**Test data / coverage**
An idea around system structure. No test available.

**Additional context**
This is another forward dependency of #14087
If this idea has certain advanced nature that is worth to introduce, I'm ask for help from someone who is experienced in stm32 development to implement that function on stm32_common and stm32h7 platforms.

### CODE REVIEW

**Comment by SalimTerryLi** [comment]
Addition:
I'd rather like to have is function implemented within `px4_arch_adc_sample`, to guarantee its synchronization. That is:
`px4_arch_adc_sample` should return a structure which contains adc sample count and vref.

Why ADC should report vref instead of lsb voltage:
If we face with lsb then float error should be considered. But vref would not be that annoying.
Some boards using 0.0004394531 as DN_to_V, but actually it has a 12bit ADC with 1.8v as vref. Similarly, 0.0008 could be boards with 12bit ADC but 3.3v as vref. The true DN_to_V value is not that perfect but vref should be a representable number and DN_full_count is already provided. Let upper function makes a division would be better, I think.

**Comment by SalimTerryLi** [comment]
I have done it through the second way. It is able to merge as currently it has no effect on anything. The actual vref can be provided in any time. Before it is provided, hard-coded value should work. But, I still need some conformation about default vref value of each boards, which is marked with TODO.

**Comment by SalimTerryLi** [comment]
I have done it through the second way. It is able to merge as currently it has no effect on anything. The actual vref can be provided in any time. Before it is provided, hard-coded value should work. But, I still need some conformation about default vref value of each boards, which is marked with TODO.

**Comment by SalimTerryLi** [comment]
Back to the first commit again.

File: platforms/nuttx/src/px4/nxp/imxrt/adc/adc.cpp [hunk 1]
```diff
@@ -186,6 +186,12 @@ uint32_t px4_arch_adc_sample(uint32_t base_address, unsigned channel)
 	result  = rR0(base_address);
 	return result;
 }
+
+float px4_arch_adc_reference_v()
+{
+	return 3.3f;	// TODO: provide true vref
```
Comment (davids5) [inline platforms/nuttx/src/px4/nxp/imxrt/adc/adc.cpp]: Please define 
BOARD_ADC_POS_REF_V 3.3 in board_common.h.  Then use that define for all of these in every adc driver. Then in the board_config.h for boards that do not have a 3.3 ref Add the define there to override the default. 

create a devcollapse the #defines in all the  in  use the define in board_config.h - It sets the wiring.

File: src/drivers/drv_adc.h [hunk 2]
```diff
@@ -82,6 +82,13 @@ void px4_arch_adc_uninit(uint32_t base_address);
  */
 uint32_t px4_arch_adc_sample(uint32_t base_address, unsigned channel);
 
+/**
+ * Get the overall reference voltage
```
Comment (davids5) [inline src/drivers/drv_adc.h]:
```suggestion
 * Get the ADC positive reference voltage
```

File: src/drivers/drv_adc.h [hunk 3]
```diff
@@ -82,6 +82,13 @@ void px4_arch_adc_uninit(uint32_t base_address);
...
+ * Currently assume that all ADC channels share the same vref.
```
Comment (davids5) [inline src/drivers/drv_adc.h]:
```suggestion
 * N.B This assume that all ADC channels share the same vref.
```

File: platforms/nuttx/src/px4/nxp/imxrt/adc/adc.cpp (hunk 1 above)
Comment (SalimTerryLi) [reply platforms/nuttx/src/px4/nxp/imxrt/adc/adc.cpp]: Done with 047f400c90

**Comment by SalimTerryLi** [comment]
> @SalimTerryLi - can we kill off?
> 
> ```
> #ifndef BOARD_ADC_POS_REF_V_FOR_CURRENT_CHAN
> #define BOARD_ADC_POS_REF_V_FOR_CURRENT_CHAN (3.3f) // Reference voltage for reading out the current channel	#define BOARD_ADC_POS_REF_V_FOR_CURRENT_CHAN (3.3f) // Reference voltage for reading out the current channel
> #endif	#endif
> #ifndef BOARD_ADC_POS_REF_V_FOR_VOLTAGE_CHAN	#ifndef BOARD_ADC_POS_REF_V_FOR_VOLTAGE_CHAN
> #define BOARD_ADC_POS_REF_V_FOR_VOLTAGE_CHAN (3.3f) // Reference voltage for reading out the voltage channel	#define BOARD_ADC_POS_REF_V_FOR_VOLTAGE_CHAN (3.3f) // Reference voltage for reading out the voltage channel
> #endif	#endif
> ```

I'll do that in the next PR #14087
